lib/ex_aws/sns/client.ex

```diff
@@ -0,0 +1,149 @@
+defmodule ExAws.SNS.Client do
+  use Behaviour
+
+  @moduledoc """
+  Defines an SNS Client
+
+  By default you can use ExAws.SNS
+
+  ## Usage
+
+  ```
+  defmodule MyApp.SNS do
+    use ExAws.SNS.Client, otp_app: :my_otp_app
+  end
+  ```
+
+  In your config:
+
+  ```
+  config :my_otp_app, :ex_aws,
+    sns: [], # SNS config goes here
+  ```
+
+  You can now use MyApp.SNS as the root module for the SNS API without
+  needing to pass in a particular configuration. This enables different OTP
+  apps to configure their AWS configurations separately.
+
+  The alignment with a particular OTP app while convenient is however entirely
+  optional.
+
+  The following also works:
+
+  ```
+  defmodule MyApp.SNS do
+    use ExAws.SNS.Client
+
+    def config_root do
+      Application.get_all_env(:my_aws_config_root)
+    end
+  end
+  ```
+
+  ExAws now expects the config for that client to live under `:my_aws_config_root`:
+
+  ```elixir
+  config :my_aws_config_root
+    sns: [] # SNS config goes here
+  ```
+
+  Default config values can be found in ExAws.Config.
+
+  ## General notes
+
+  TODO
+
+  ## Examples
+
+  TODO
+
+  http://docs.aws.amazon.com/sns/latest/APIReference/API_Operations.html
+  """
+
+  ## Topics
+  ######################
+
+  @type topic_name :: binary
+  @type topic_arn  :: binary
+  @type topic_attribute_name ::
+    :policy |
+    :display_name |
+    :delivery_policy
+
+  @doc "List topics"
+  defcallback list_topics() :: ExAws.Request.response_t
+  defcallback list_topics(opts :: [next_token: binary]) :: ExAws.Request.response_t
+
+  @doc "Create topic"
+  defcallback create_topic(topic_name :: topic_name) :: ExAws.Request.response_t
+
+  @doc "Get topic attributes"
+  defcallback get_topic_attributes(topic_arn :: topic_arn) :: ExAws.Request.response_t
+
+  @doc "Set topic attributes"
+  defcallback set_topic_attributes(attribute_name :: topic_attribute_name,
+                                   attribute_value :: binary,
+                                   topic_arn :: topic_arn) :: ExAws.Request.response_t
+
+  @doc "Delete topic"
+  defcallback delete_topic(topic_arn :: topic_arn) :: ExAws.Request.response_t
+
+  ## Publishing
+  ######################
+
+  @type message_attribute :: %{
+    :name => binary,
+    :data_type => :string | :number | :binary,
+    :value => {:string, binary} | {:binary, binary}
+  }
```

Comment:
@bruce let me know what you think about this typespec for message attributes. While keyword lists I think are appropriate for function options, a message attribute is really an entity in and of itself, which I prefer to use maps for. Additionally since you can have several of them, you end up with a list of keyword lists which can be visually confusing.